### PR TITLE
Remove nonreentrantlock type

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This type was created and used in a time that predated both TPL and helper types like SemaphoreSlim.  We can remove this and just use what the platform gives us for this basic concept.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EventMap.cs [hunk 1]
```diff
@@ -76,7 +77,7 @@ private ImmutableArray<Registry<TEventHandler>> GetRegistries<TEventHandler>(str
         private ImmutableArray<Registry<TEventHandler>> GetRegistries_NoLock<TEventHandler>(string eventName)
             where TEventHandler : class
         {
-            _guard.AssertHasLock();
+            Contract.ThrowIfTrue(_guard.CurrentCount > 0);
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EventMap.cs:80]: ❗ I'm concerned about the change in semantics here

**Review by sharwell — DISMISSED** [review]
Requesting changes so we can discuss ways to preserve semantics

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EventMap.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EventMap.cs:80]: it's a weaker check.  but will still at least fire when very wrong.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs [hunk 2]
```diff
@@ -59,7 +59,7 @@ internal sealed class AsyncLazy<T>
         /// by using a single lock for all AsyncLazy instances.  Only trivial and non-reentrant work
         /// should be done while holding the lock.
         /// </summary>
-        private static readonly NonReentrantLock s_gate = new(useThisInstanceForSynchronization: true);
+        private static readonly SemaphoreSlim s_gate = new(initialCount: 1);
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs:62]: 📝 This one has a mix of cancellation tokens and not

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EventMap.cs [hunk 3]
```diff
@@ -6,13 +6,14 @@
...
 using System.Linq;
+using System.Threading;
 using Microsoft.CodeAnalysis.ErrorReporting;
 
 namespace Roslyn.Utilities
 {
     internal class EventMap
     {
-        private readonly NonReentrantLock _guard = new();
+        private readonly SemaphoreSlim _guard = new(initialCount: 1);
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EventMap.cs:16]: 💡 This one doesn't use cancellation tokens, but it does include specific assertions about holding a lock. It seems like a good candidate for `object`/`Monitor`.

**Review by sharwell — COMMENTED** [review]
Overall, I do feel a bit concerned about the change in semantics with regard to deadlock prevention. There are a few cases called out where cancellation tokens are never used, and as such can change from `NonReentrantLock` to `object` without any change in behavior.

**Comment by CyrusNajmabadi** [comment]
I would prefer not changing to object as it then requires ensuring that all callsites add throws that they don't already have teh lock.  It would be easy to miss, and then end up accidentally reentering.

I'm not very concerned about deadlocking.  If things are broken, having it deadlock is just as broken.  Practically speaking, we haven't even hit this exception well... ever.  So going through contortions to try to preserve it seems unnecessary.  We have tons of code using semaphoreslim effectively, and it was never a problem that that code would similarly deadlock, so we should just keep things in that form for the rest of our code.

**Comment by jasonmalinowski** [comment]
I'm not strongly against a long-term removal of this but I do want to correct this one statement:

> Practically speaking, we haven't even hit this exception well... ever.

Right now there's a bug against me where we are hitting this and this PR is reminding me I forgot to fix that for 17.10 Preview 1. And a very quick Kusto search implies we may have at least two other bugs too where we'd be replacing an exception with a hang.  There's an internal outage of some tools right now I'd use to file bugs so I'll follow up on those others when it's back, but in the mean time I'd least request that we don't merge this until those fixes go in, or we better understand those other ones to deem them not a significant issue.

But it does make me hope that if we are going ahead with the new C# lock feature that the framework would have a non-reentrant compatible lock with it and then we could ditch this code and just use that.

**Comment by CyrusNajmabadi** [comment]
> but in the mean time I'd least request that we don't merge this until those fixes go in,

I can deal with that.

**Comment by ToddGrun** [comment]
@jasonmalinowski -- Can you share the query with how these exceptions are being found? Is it possible that we are under-reporting the number of times this occurs by either blanket catching or losing the exception otherwise?

**Pull request closed without merging**